src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1793 @@
+use core::convert::TryInto;
+use core::iter::FromIterator;
+use core::ops::{Index, IndexMut};
+use crate::domtree::dominates;
+use crate::{domtree, postorder, AllocationKind, Block, Inst, InstPosition, Operand, OperandConstraint, OperandKind, OperandPos, PReg, PRegSet, RegClass, SpillSlot, VReg};
+use crate::{Function, MachineEnv, ssa::validate_ssa, ProgPoint, Edit, Output};
+use crate::{cfg::CFGInfo, RegAllocError, Allocation, ion::Stats};
+use alloc::collections::{BTreeSet, VecDeque};
+use alloc::vec::Vec;
+use hashbrown::{HashSet, HashMap};
+
+use std::println;
+use std::format;
+
+mod lru;
+mod iter;
+use lru::*;
+use iter::*;
+
+#[derive(Debug)]
+struct Allocs {
+    allocs: Vec<Allocation>,
+    /// `inst_alloc_offsets[i]` is the offset into `allocs` for the allocations of
+    /// instruction `i`'s operands.
+    inst_alloc_offsets: Vec<u32>,
+}
+
+impl Allocs {
+    fn new<F: Function>(func: &F, env: &MachineEnv) -> Self {
+        // The number of operands is <= number of virtual registers
+        // It can be lesser in the case where virtual registers are used multiple
+        // times in a single instruction.
+        let mut allocs = Vec::with_capacity(func.num_vregs());
+        let mut inst_alloc_offsets = Vec::with_capacity(func.num_vregs());
+        for inst in 0..func.num_insts() {
+            let operands_len = func.inst_operands(Inst::new(inst)).len() as u32;
+            inst_alloc_offsets.push(allocs.len() as u32);
+            for _ in 0..operands_len {
+                allocs.push(Allocation::none());
+            }
+        }
+        Self {
+            allocs,
+            inst_alloc_offsets,
+        }
+    }
+}
+
+impl Index<(usize, usize)> for Allocs {
+    type Output = Allocation;
+
+    /// Retrieve the allocation for operand `idx.1` at instruction `idx.0`
+    fn index(&self, idx: (usize, usize)) -> &Allocation {
+        &self.allocs[self.inst_alloc_offsets[idx.0] as usize + idx.1]
+    }
+}
+
+impl IndexMut<(usize, usize)> for Allocs {
+    fn index_mut(&mut self, idx: (usize, usize)) -> &mut Allocation {
+        &mut self.allocs[self.inst_alloc_offsets[idx.0] as usize + idx.1]
+    }
+}
+
+#[derive(Debug)]
+pub struct Env<'a, F: Function> {
+    func: &'a F,
+
+    /// The current allocations for all virtual registers.
+    vreg_allocs: Vec<Allocation>,
+    /// Spillslots for all virtual registers.
+    /// `vreg_spillslots[i]` is the spillslot for virtual register `i`.
+    vreg_spillslots: Vec<SpillSlot>,
+    /// The virtual registers that are currently live.
+    live_vregs: HashSet<VReg>,
+    /// Allocatable free physical registers for classes Int, Float, and Vector, respectively.
+    freepregs: PartedByRegClass<BTreeSet<PReg>>,
```

Comment:
as Amanieu mentioned earlier, a `PRegSet` should be much more efficient here (and will allow us to replace some iteration below with intersection/union operations I think).